### ISSUE TITLE
Ctrl + [ doesn't function as Escape in helix 23.03

### ISSUE THREAD

**Issue by xcpky** [helix-editor/helix#6551]
I think this is because of the new enhanced keyboard protocol enabled in 23.03. Is there a way to disable it? Although I can remap Ctrl+[ to mode change command, escape has more functions than only mode change. And there is no way to remap a key combination to a specific keycode in helix, which is available in vim.

**Comment by the-mikedavis** [comment]
Yeah, the enhanced keyboard protocol change was breaking if you are relying on ambiguous keycodes like `C-[`. IMO you should not rely on these ambiguous keycodes because it's essentially coincidence that they worked in the first place. And I don't like adding a config option for this since I think the use-case is very niche. Ideally this should be fixed by allowing remapping Escape in the remaining places where it is hard-coded (#5505).

On the other hand though, this could be covered by a config option which overrides the detection, `true` to assume the protocol is supported and `false` to assume not. (This has been discussed previously but I can't find the issue.)

**Comment by wvwwvwwv** [comment]
`Ctrl + [` is the de facto standard `esc` alternative on macOS, and I don't think the usage is niche, because `Ctrl + [` is especially useful when your keyboard does not have an `esc` key (which is true for many Mac/iPad users). I'm really looking forward to the upcoming `esc` remapping feature.

**Comment by mk-f** [comment]
I also don't think that using `ctrl+[` as `esc` is a niche use-case, on keyboards with us-layout it is more comfortable to reach (both pinkies) than the actual `esc`.

For what it's worth you can run hx inside tmux for the moment as a workaround.

**Comment by jaese** [comment]
These two other [control character mappings](https://en.wikipedia.org/wiki/Control_character#How_control_characters_map_to_keyboards), backspace `ctrl-h` and line feed `ctrl-j`, are still supported after the enhanced keyboard protocol change, so I don't think it is a stretch to keep supporting `ctrl-[` as escape by default.

**Comment by n8henrie** [comment]
With neither `ctrl-c` and `ctrl-[` working for this purpose, I'm left having to leave my home row to exit insert mode (via `Esc`). Is this by design, or am I missing a recommended alternative?

**Comment by pascalkuthe** [comment]
you can always rebind the keymapping in your own config.

**Comment by sdemura** [comment]
@pascalkuthe can you provide an example?

```
❯ hx
Bad config: TOML parse error at line 19, column 9
   |
19 | "C-[" = "esc"
   |         ^^^^^
No command named 'esc
```

**Comment by soifou** [comment]
@sdemura, almost :) what you want is: 

```toml
[keys.insert]
  "C-[" = "normal_mode"
```

**Comment by xcpky** [comment]
> 

but this workaround can't substitute all the esc usage with 'Ctrl-[', so a feature of keycode remapping should be the best solution.

**Comment by samuelstevens** [comment]
```
[keys.insert]
"C-[" = "normal_mode"
```

This doesn't work for me with `helix 23.05 (7f5940be)` on Ubuntu 22.04 in tmux. It works outside of tmux, but not inside. Any advice?

**Comment by pascalkuthe** [comment]
Termux does not support the kitty enhanced keyboard protocol which allows disambiguation of these keycodes. Since terminal multiplexers essentially function as a terminal proxy they must support all of the features of the host terminal (that the user wants to use)

**Comment by samuelstevens** [comment]
If tmux doesn't support the disambiguation of ctrl-[ and esc, then shouldn't ctrl-[ work as esc in helix? or am I misunderstanding something?

**Comment by the-mikedavis** [comment]
You can check whether it's enabled by checking the log when running in verbose mode: https://github.com/helix-editor/helix/wiki/Terminal-Support#enhanced-keyboard-protocol

It sounds to me like tmux is probably eating the keycode.

**Comment by teancom** [comment]
Came here as someone who has been evangelizing mapping CAPS LOCK to CTRL and then using ctrl-[ in vim for over 20 years now. Specifically as someone who has RSI issues and is able to use caps_lock-[ with two very subtle finger movements vs. any alternative. If it was a mere accident that that worked, it's an accident that was true for every combination of vi/vim/nvim/etc. in every terminal emulator and every operating system I've ever used.

**Comment by kj** [comment]
Yes I agree, I think all of this talk of `C-[` being an accident is a little dismissive. Perhaps it was an 'accident' once but its usage is certainly not niche (as this issue proves). The reality is that many people expect and rely on this behaviour. This just hit me today as Alacritty 0.13.0 now implements the kitty keyboard protocol.

**Comment by geo-ant** [comment]
Hi, just came here to say hat I love helix, and wanted to voice my support that mapping control to caps lock and using C-[ is not a niche use case and should please have first class support. Since I am just some guy on the internet, I'll try to supplement my argument with some appeals to authority in hopes that this rhetorical flourish will convince people :) 

* In his book Practical Vim, p.30 author Drew Neil recommends mapping caps lock to ctrl and using C-[ as escape. Quote: "The C-[ mapping is synonymous with Esc , and it’s easier to type when the Ctrl key is within easy reach."
* In his talk [Stop Writing Dead Programs](https://www.youtube.com/watch?v=8Ab3ArE8W3s) around the 18:00 minute mark, Jack Rusher says he still maps caps lock to control because it makes his life easier on unix machines (and gives an interesting historical reason for that :) )

I can also try and hop on to help with the implementation if this is a resource issue.

**Comment by kj** [comment]
In case it helps anyone… until Helix supports this, using [keyd](https://github.com/rvaiya/keyd) to map `C-[` to escape seems to work.

```ini
[control]
[ = esc
```

**Comment by webdev23** [comment]
There is a couple of settings in `~/.tmux.conf` that affects the behavior of the escape sequence, and may help with this issue.

In particular the following:

`set -sg escape-time 0`

---

In Helix, found the following mapping to be particularly efficient: **Ctrl + space**
```
[keys.insert]                                                             
C-space = "normal_mode"    

[keys.normal]
C-space = "insert_mode" 

[keys.select]
C-space = "normal_mode"     
```

**Comment by wrl** [comment]
My `C-[` muscle memory is so strong that I also do it at prompts and in pickers, which are not currently re-mappable in Helix (that I can find, at least). Would be nice to have it mapped there too.

**Comment by webdev23** [comment]
To advocate for `Ctrl + [` users. 

They are in their right to notice, this is a very old behavior in VT100 terminal that `Ctrl+ [` does produce the escape sequence. 

To verify that, we may simply use `cat`:

`cat`
We press Escape
That does print `^[`
Now we press `Ctrl + [`
That does also print `^[`

Further documentations can be found with keywords: **control-escape sequence VT100**

I do personally, as a non-Qwerty keyboard user, have to use the key combo `Ctrl + AltGr + 4` to output `Ctrl + [`. 

However I do warmly recommend that Helix supports `Ctrl + [` as the escape sequence due to legacy reasons.

**Comment by zhaob1n** [comment]
hi everyone, I have found a solution for kitty but forgot to share here. For kitty users, add the following to kitty.conf.
```
map ctrl+[ send_text kitty \x1b
```

For other terminals, there might be similar solutions.

**Comment by miguelweber** [comment]
For alacritty, add this to your alacritty.toml file:

```
[[keyboard.bindings]]
key = "["
mods = "Control"
chars = "\u001b"
```

**Comment by Griffin-Brome** [comment]
For ghostty:

```
keybind = ctrl+left_bracket=text:\x1b
```

Credit to https://github.com/ghostty-org/ghostty/discussions/4764

**Comment by Shad0wRim** [comment]
> ```
> map ctrl+[ send_text kitty \x1b
> ```

For kitty, send_key allows the remapping of ctrl+j to enter as well
```
map ctrl+[ send_key escape
map ctrl+j send_key enter
```
